### PR TITLE
signal: use work_cancel_sync() to fix used after free

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary

wqueue: add work_cancel_sync() support
signal: use work_cancel_sync() to fix used after free

    user thread:                             hpwork:
    timer_create() with SIGEV_THREAD
    timer_settime()
        irq -> work_queue()                  add nxsig_notification_worker to Q
    timer_delete()
        nxsig_cancel_notification()
                                             call nxsig_notification_worker()
        work_cancel()
        timer_free()
                                             nxsig_notification_worker() used after free
    
    root cause:
    work_cancel() can't cancel work completely, the worker may alreay be running.
    
    resolve:
    use work_cancel_sync() API to cancel the work completely

## Impact

work queue

## Testing

BES board & SIM

### CODE REVIEW

File: sched/wqueue/kwork_cancel.c [hunk 1]
```diff
@@ -93,6 +96,21 @@ static int work_qcancel(FAR struct kwork_wqueue_s *wqueue,
...
       ret = OK;
     }
+  else if (nthread > 0)
+    {
+      int wndx;
+
+      for (wndx = 0; wndx < nthread; wndx++)
+        {
+          if (wqueue->worker[wndx].work == work &&
+              wqueue->worker[wndx].pid != _SCHED_GETTID())
```
Comment (pkarashchenko) [inline sched/wqueue/kwork_cancel.c:106]: Why _SCHED_GETTID macro is used? It is not a libc call, so we should call nx call directly here

**Comment by pkarashchenko** [comment]
Please rebase PR to fix CI failure

File: sched/wqueue/kwork_cancel.c (hunk 1 above)
Comment (xiaoxiang781216) [reply sched/wqueue/kwork_cancel.c:106]: since we want to avoid sched note syscall hook catch the kernel internal activities.

Comment (pkarashchenko) [reply sched/wqueue/kwork_cancel.c:106]: Still not clear why we can't use `nxsched_gettid()` here. Could you please provide more detailed description?

Comment (xiaoxiang781216) [reply sched/wqueue/kwork_cancel.c:106]: Please see this patch: https://github.com/apache/nuttx/pull/8394

Comment (pkarashchenko) [reply sched/wqueue/kwork_cancel.c:106]: Yes, exactly. The `_SCHED_GETTID` is used in the places where code is shared between apps and kernel, like `libc`, so the question is do we compile code for kworker out of kernel?

Comment (GUIDINGLI) [reply sched/wqueue/kwork_cancel.c]: change to nx API

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**